train_runner.py

```diff
@@ -7,71 +7,115 @@
+    PRINT_INTERVAL = 100
+
+    # TODO: for local use
+    DATA_PATH = MODELS_PATH = os.path.abspath(__file__+'/../')
+
+    # Data
+    DATASET_NAME = 'IMDB'
+    TRAIN_BATCH_SIZE = 16
+    TEST_BATCH_SIZE = 16
+    TRAIN_SIZE = 20000 # For debug
+    # maximum length of allowed sentence - can be also None
+    MAX_LEN = None
+
+    # Transformer model
+    N_LAYERS = 6
+    H_DIM = 512
+    N_ATTN_HEAD = 8
+    FC_DIM = 2048
+    DO_RATE = 0.1
+
+    # Classification model
+    N_STYLES = 2
+    DO_RATE_CLS = 0.5
+
+    # Train
+    N_EPOCHS = 100
+    PATIENCE = 3
+    LR = 0.1
+    OPT_WARMUP_FACTOR = 4000
+    REC_LAMBDA = 0.5
+    CLS_STEPS = 500
+    TRANS_STEPS = 500
+
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+    def pprint(self):
+        print('Params for experiment:')
+        for attr in dir(self):
+            if attr.startswith('_'):
+                pass
+            else:
+                print("%s = %r" % (attr, getattr(self, attr)))
+
+params = Params()
+params.pprint()
+
+TEXT, word_embeddings, train_iter, test_iter = load_dataset(data_source=params.DATASET_NAME, fix_length=params.MAX_LEN,
+                                                            device=params.device, batch_size_train=params.TRAIN_BATCH_SIZE,
+                                                            batch_size_test=params.TEST_BATCH_SIZE)
+
+print('Train dataset len: {} Test dataset len: {}'.format(len(train_iter.dataset), len(test_iter.dataset)))
 
 # Clear CUDA memory if needed
 # TODO: local use
 # torch.cuda.empty_cache()
 
 ### Init models ###
-device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
-print('Device is {}'.format(device))
 
-vocab_size = len(tokenizer.vocab)
-rm = ResourcesManager(vocab_size, MODELS_PATH, "basic", device, load_path=MODELS_PATH)
+vocab_size = len(TEXT.vocab)
+rm = ResourcesManager(vocab_size, params.MODELS_PATH, "basic_exp_{}".format(params.EXP_NUM),
+                      params, load_path=params.MODELS_PATH)
 models = rm.get_models()
 model_enc = models["enc"]
 model_dec = models["dec"]
 model_cls = models["cls"]
 
+# TODO: if using this option set H_DIM = 300
+# model_enc = load_pretrained_embedding_to_encoder(model_enc, word_embeddings)
+
 ### Init losses ###
 cls_criteria = nn.CrossEntropyLoss()
 seq2seq_criteria = LabelSmoothing(size=vocab_size, padding_idx=0)
 ent_criteria = EntropyLoss()
 
 ### Init optimizers ###
-cls_opt = NoamOpt(H_DIM, 2, OPT_WARMUP_FACTOR,
-            torch.optim.Adam(model_cls.parameters(), lr=LR, betas=(0.9, 0.98), eps=1e-9))
-enc_opt = get_std_opt(model_enc)
-dec_opt = get_std_opt(model_dec)
+cls_opt = get_std_opt(model_cls, params)
+enc_opt = get_std_opt(model_enc, params)
+dec_opt = get_std_opt(model_dec, params)
 
-model_enc.train()
-model_dec.train()
-model_cls.train()
+early_stop = EarlyStopping(params.PATIENCE)
+for epoch in range(Params.N_EPOCHS):
+  rm.on_epoch_end(epoch)
```

Comment:
shouldn't this happen on end, not beginning?

Reply:
True, used of debug